### DISCUSSION TITLE
Use `x_dot` kwarg in `model.fit` in PDEs with `WeakPDELibrary()`

### DISCUSSION THREAD

**Discussion by georgemilosh** [dynamicslab/pysindy#444] [Q&A]
So the problem occurs when I try to compute `x_dot` myself and use it in PDEs. Here is my code 

```
import numpy as np
import pysindy as ps
from scipy.io import loadmat

data = loadmat("../../data/kuramoto_sivishinky.mat")
time = np.ravel(data["tt"])
x = np.ravel(data["x"])
u = data["uu"]
dt = time[1] - time[0]
u_dot = ps.FiniteDifference(axis=1)._differentiate(u, t=dt)
u = u.reshape(len(x), len(time), 1)
X, T = np.meshgrid(x, time)
XT = np.asarray([X, T]).T
#pde_lib = ps.WeakPDELibrary(spatiotemporal_grid=XT,)
library_functions = [lambda x: x, lambda x: x * x]
library_function_names = [lambda x: x, lambda x: x + x]
pde_lib = ps.WeakPDELibrary(
    library_functions=library_functions,
    function_names=library_function_names,
    derivative_order=4,
    spatiotemporal_grid=XT,
    is_uniform=True,
    K=200,
    include_interaction=True,
    periodic=True,
)

optimizer = ps.STLSQ()
threshold = 2
optimizer = ps.STLSQ(threshold=threshold, alpha=0.05, normalize_columns=True)
model = ps.SINDy(feature_library=pde_lib, optimizer=optimizer)
model.fit(u, x_dot =u_dot)
model.print()
```

I get 
```
ValueError: operands could not be broadcast together with shapes (200,) (1024,) 
```

If I try to flatten u_dot I get
```
ValueError: operands could not be broadcast together with shapes (200,) (257024,) 
```

I don't get this error if I work with standard kwargs:

```
import numpy as np
import pysindy as ps
from scipy.io import loadmat

data = loadmat("../../data/kuramoto_sivishinky.mat")
time = np.ravel(data["tt"])
x = np.ravel(data["x"])
u = data["uu"]
dt = time[1] - time[0]
u_dot = ps.FiniteDifference(axis=1)._differentiate(u, t=dt)
u = u.reshape(len(x), len(time), 1)
X, T = np.meshgrid(x, time)
XT = np.asarray([X, T]).T
pde_lib = ps.WeakPDELibrary(spatiotemporal_grid=XT,)

optimizer = ps.STLSQ()
model = ps.SINDy()
model.fit(u, x_dot=u_dot.flatten())
model.print()
```

Here flatten u helps. So how can I provide my own x_dot info for PDEs?

My package version: 
0.1.dev1618+g694c904.d20231219 3.8.18 | packaged by conda-forge | (default, Oct 10 2023, 15:44:36) 
[GCC 12.3.0]

**Comment by Jacob-Stevens-Haas** [comment]
`WeakPDELibrary` doesn't even use the derivative - even if it did accept `x_dot` without error, it would be ignored.  A good argument for #351.

Does it work for `PDELibrary`?

**Reply by georgemilosh** [reply]
So it seems that the `PDELibrary` does not throw an error, although I have to manually flatten u_dot:

```
import numpy as np
from scipy.io import loadmat
from sklearn.metrics import mean_squared_error

import pysindy as ps

# Ignore matplotlib deprecation warnings
import warnings

warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

# Seed the random number generators for reproducibility
np.random.seed(100)

data = loadmat("data/burgers.mat")
time = np.ravel(data["t"])
x = np.ravel(data["x"])
u = np.real(data["usol"])
dt = time[1] - time[0]
dx = x[1] - x[0]
rmse = mean_squared_error(u, np.zeros(u.shape), squared=False)
# add 20% noise (note the impact on derivatives depends on step size...)
u = u + np.random.normal(0, rmse / 5.0, u.shape)


u = np.reshape(u, (len(x), len(time), 1))

# Define weak form PDE library
library_functions = [lambda x: x, lambda x: x * x]
library_function_names = [lambda x: x, lambda x: x + x]

pde_lib = ps.PDELibrary(
    library_functions=library_functions,
    function_names=library_function_names,
    derivative_order=2,
    spatial_grid=x,
    include_bias=False,
    is_uniform=True,
)

# Fit and predict with the non-weak model
opt = ps.SR3(
    threshold=0.05, thresholder="l0", tol=1e-10, normalize_columns=True, max_iter=1000
)
model_for_prediction = ps.SINDy(feature_library=pde_lib, optimizer=opt)
#model_for_prediction.fit(u, t=dt) # default behavior, gives:
## (x0)' = 0.011 x0 + -0.031 x0x0 + -0.001 x0_1 + -0.012 x0x0_1 + -0.028 x0x0x0_1
## I had to add t=dt as an argument to match the result below

#model_for_prediction.fit(u, x_dot=u_dot) # doesn't work!

u_dot = ps.FiniteDifference(axis=1)._differentiate(u, t=dt)
model_for_prediction.fit(u, x_dot=u_dot.flatten(),t =dt) # my guess. Produces 
# (x0)' = 0.108 x0 + -0.313 x0x0 + -0.012 x0_1 + -0.117 x0x0_1 + -0.278 x0x0x0_1 + -0.004 x0x0_11 + 0.005 x0x0x0_11


model_for_prediction.print()
```

**Reply by Jacob-Stevens-Haas** [reply]
Thanks for troubleshooting this with me.  Internally, instead of `flatten()`, we use `pysindy.utils.concat_sample_axis(x_dot)`.  But before that can work, `x_dot` needs  its axes labeled.

As your solution suggests, we have a convention that all but the final axis gets turned into samples for the regression.  We apply this convention early so that internal code refers to the labels rather than convention (e.g. `FiniteDifference(axis=x.ax_time)` instead of `FiniteDifference(axis=-2)`).  This makes it easier to understand a lot of code in the feature libraries.  It does lead to issues like yours, however, when we don't wrap user inputs.  You can access this convention `pysindy.utils.comprehend_axes` and label them with `pysindy.AxesArray`:

```python
u = ps.AxesArray(u, ps.utils.comprehend_axes(u))
u_dot = ps.FiniteDifference(axis=u.ax_time)(u, t=dt)
model_for_prediction.fit(u, x_dot=u_dot,t =dt)
```
^--Note how, once `u` is labeled with axes, so is `u_dot`
^--P.S. - `FiniteDifference._differentiate()` is not part of the public API, but the `FiniteDifference` object itself is callable.


You can look at the output of `comprehend_axes()` to see what `AxesArray` expects in that second argument.  We're working to make `AxesArray` more stable for public use (pushed most recent commit yesterday), so next release will be _slightly_ backwards incompatible (just modifying the defaults to make it more generic).

Regardless, we do want to support the calling signature that you used, so this is an actual bug.
`

**Reply by georgemilosh** [reply]
Thanks Jacob. 
So the reason i wanted to overwrite `x_dot` was because I wanted to try `WeakPDELibrary` in the example below (or something similar). The example comes from the end of `pysindy/examples/1_feature_overview/example.py`
```
import numpy as np
import pysindy as ps
# Define the spatial grid

if __name__ != "testing":
    nx = 50
    ny = 100
else:
    nx = 6  # must be even
    ny = 10
Lx = 1
Ly = 1
x = np.linspace(0, Lx, nx)
dx = x[1] - x[0]
y = np.linspace(0, Ly, ny)
dy = y[1] - y[0]
X, Y = np.meshgrid(x, y, indexing="ij")

# Define rho
rho = X**2 + Y**2

# Generate the PDE data for phi by fourier transforms
# since this is homogeneous PDE
# and we assume periodic boundary conditions
nx2 = int(nx / 2)
ny2 = int(ny / 2)
# Define Fourier wavevectors (kx, ky)
kx = (2 * np.pi / Lx) * np.hstack(
    (np.linspace(0, nx2 - 1, nx2), np.linspace(-nx2, -1, nx2))
)
ky = (2 * np.pi / Ly) * np.hstack(
    (np.linspace(0, ny2 - 1, ny2), np.linspace(-ny2, -1, ny2))
)

# Get 2D mesh in (kx, ky)
KX, KY = np.meshgrid(kx, ky, indexing="ij")
K2 = KX**2 + KY**2
K2[0, 0] = 1e-5

# Generate phi data by solving the PDE
phi = np.real(np.fft.ifft2(-np.fft.fft2(rho) / K2))

# Make del^2 phi
phi_xx = ps.FiniteDifference(d=2, axis=0)._differentiate(phi, dx)
phi_yy = ps.FiniteDifference(d=2, axis=1)._differentiate(phi, dy)

# Define a PolynomialLibrary, FourierLibrary, and PDELibrary
poly_library = ps.PolynomialLibrary(include_bias=False)
fourier_library = ps.FourierLibrary()
X_mesh, Y_mesh = np.meshgrid(x, y)
pde_library = ps.PDELibrary(
    derivative_order=1, spatial_grid=np.asarray([X_mesh, Y_mesh]).T
)

# Inputs are going to be all the variables [phi, X, Y].
# Remember we can use a subset of these input variables to generate each library
data = np.transpose(np.asarray([phi, X, Y]), [1, 2, 0])

# The 'x_dot' terms will be [phi_xx, X, Y]
# Remember these are the things that are being fit in the SINDy regression
Laplacian_phi = phi_xx + phi_yy + abs(np.min(phi_xx + phi_yy))
data_dot = np.transpose(np.asarray([Laplacian_phi, X, Y]), [1, 2, 0])

# Tensor polynomial library with the PDE library
tensor_array = [[1, 0, 1]]

# Remove X and Y from PDE library terms because why would we take these derivatives
inputs_per_library = [(0, 1, 2), (0, 1, 2), (0,)]

# Fit a generalized library of 3 feature libraries + 1 internally
# generated tensored library and only use the input variable phi
# for the PDELibrary. Note that this holds true both for the
# individual PDELibrary and any tensored libraries constructed from it.
generalized_library = ps.GeneralizedLibrary(
    [poly_library, fourier_library, pde_library],
    tensor_array=tensor_array,
    inputs_per_library=inputs_per_library,
)
optimizer = ps.STLSQ(threshold=8, alpha=1e-3, normalize_columns=True)
model = ps.SINDy(feature_library=generalized_library, optimizer=optimizer)
model.fit(data, x_dot=data_dot)

# Note scale of phi is large so some coefficients >> 1
# --> would want to rescale phi with eps_0 for a harder problem
model.print()
```
So making some changes here to go from `PDELibrary` to `WeakPDELibrary` is not possible?

**Reply by Jacob-Stevens-Haas** [reply]
Correct - it is not possible.  `WeakPDELibrary` is a bit of a misnomer - it implies weak terms across the equation, not just the feature library.  This could theoretically be allowed with #351 depending upon how it would be implemented.

**Reply by georgemilosh** [reply]
I see. So if I want to discover residual terms of a PDE with mostly known terms how would I proceed then? 
$$\dot y = f(known) + g(unknown)$$
I was planning to overwrite `x_dot` with already known term:
$$\dot y \rightarrow \dot y - f(known)$$
But this is not currently possible with `WeakPDELibrary`. 

Another possibility seems to put all the known terms into one variable, so that they act as a source using SINDYc:
$$\dot y = u + g(unknown)$$
and maybe restrict the library of `u` to only contain linear term and no interaction terms between `u` and `g`? 

Would  that work? Is there some other smart way to do this in this package?

**Reply by Jacob-Stevens-Haas** [reply]
Heads up, LaTeX on github doesn't allow a space between the dollar signs and the text between them.  (I edited your post to format the equations)

In general, you need to use ConstrainedSR3 or MIOSR with equality constraints on the terms that are known.

**Reply by georgemilosh** [reply]
Thanks, I will check them out later. The decision to set time axis as the last (-1) rather than first (0) seems so counter-intuitive to me. Also when I try to apply `GridSearchCV` to these high dimensional cases (with spatial dimensions rather than time), there are errors because I think `GridSearchCV` will treat `axis=0` as sampling axis, which in case of PDEs corresponds to the spatial dimension. Has anyone tried applying `GridSearchCV` here to PDEs?

**Reply by Jacob-Stevens-Haas** [reply]
People probably have in the past.  But the time axis isn't last, it's second-to-last, and as I mentioned, it's always accessible with `arr.ax_time` (or `arr.ax_sample` once we've calculated features, which _is_ first)

I don't think `GridSearchCV` assumes anything about axes.  It's either what you're telling it, or it's what we're doing internally.

**Reply by georgemilosh** [reply]
Thanks for the useful explanations. I see you are doing some upgrades of the package which is quite welcome. What I was wondering if it will be possible to create `WeakPDELibrary()` that is fully spatial. That is to say, it would treat time coordinate as samples and not integrate over them. This would be useful in situations where time is sampled sparsely and only spatial derivatives are of interest. This is again applicable for the regression problem where l.h.s. is something else, not a time derivative.

**Reply by Jacob-Stevens-Haas** [reply]
If you want to do it yourself, you can't use `SINDy` class, but you can manually fit/transform the data with a `WeakPDELibrary`, then send it to any optimizer to solve.
